### PR TITLE
Make it clear that `StatisticsConverter` can not panic

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-rs/issues/6164

# Rationale for this change
 
As discussed with @Kev1n8 on https://github.com/apache/arrow-rs/pull/6181#discussion_r1702702575 the `DataType` is always known and provided, so having the code check for `None` is unnecessary and confusing.

# What changes are included in this PR?

Remove the `Option` handling in code that is always invoked with `Some`

# Are there any user-facing changes?
No

### CODE REVIEW

File: parquet/src/arrow/arrow_reader/statistics.rs [hunk 1]
```diff
@@ -1008,23 +1003,22 @@ macro_rules! get_data_page_statistics {
...
+                DataType::Struct(_) |
+                DataType::Union(_, _) |
+                DataType::Map(_, _) |
+                DataType::RunEndEncoded(_, _) => {
                     let len = $iterator.count();
                     // don't know how to extract statistics, so return a null array
-                    Ok(new_null_array($data_type.unwrap(), len))
+                    Ok(new_null_array($data_type, len))
                 },
-                None => unimplemented!()  // not sure how to handle this
```
Comment (alamb) [inline parquet/src/arrow/arrow_reader/statistics.rs:1027]: The point of this PR is to remove this line

File: parquet/src/arrow/arrow_reader/statistics.rs [hunk 2]
```diff
@@ -1052,10 +1046,7 @@ fn max_statistics<'a, I: Iterator<Item = Option<&'a ParquetStatistics>>>(
 
 /// Extracts the min statistics from an iterator
 /// of parquet page [`Index`]'es to an [`ArrayRef`]
-pub(crate) fn min_page_statistics<'a, I>(
-    data_type: Option<&DataType>,
-    iterator: I,
-) -> Result<ArrayRef>
+pub(crate) fn min_page_statistics<'a, I>(data_type: &DataType, iterator: I) -> Result<ArrayRef>
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/statistics.rs:1049]: This is just a drive-by observation that this will generate a fair amount of code for every iterator, it might be worth just paying the cost of just forcing `&[ParquetStatistics]` or similar

**Review by tustvold — APPROVED** [review]

File: parquet/src/arrow/arrow_reader/statistics.rs (hunk 2 above)
Comment (alamb) [reply parquet/src/arrow/arrow_reader/statistics.rs:1049]: I think you are right that  `&[ParquetStatistics]` is likely to be the most commonly used iterator, and this generic formulation may be simply over engineering

I harbor some idea that we will be able to use the iterator API to quickly extract statistics for many files at once into a single large array (e.g. by reading the data using `ParquetMetaDataReader`) so we can prune across 100s of files very fast.

However, I will fully admit that no such code exists yet (either in our InfluxDB code or as an example) 🤔

**Comment by alamb** [comment]
Thanks for the review @tustvold
